### DISCUSSION TITLE
Error - Step forward 5s (aka > right button)

### DISCUSSION THREAD

**Discussion by BloodmoonCyberFox** [iina/iina#4532] [Q&A]
Hello Developers,

I would need to seek your help, even if I have reset the settings and reinstalled the IINA.

For every time I press on the right button for 5s step forward, IINA would reset the video to 00:00 - it is the same when I press the Playback > Step Forward 5s (as shown in the captured recording - https://drive.google.com/file/d/1nrxT5PnAaXHx_NomPBFs_cK8nzA37uLE/view?usp=drive_link)

What should I do??

**Comment by low-batt** [comment]
On this:
> What should I do??

Add the video being played in the screen capture to the Goo folder.

You would think seeking to a particular spot in a video would be trivial, but it is not at all. It is dependent upon how the video is encoded.

To investigate this a developer needs the video that exhibits the problem.

**Comment by BloodmoonCyberFox** [comment]
take it or leave it

**Comment by low-batt** [comment]
I'm confused? I checked the link again but only see the screen recording?

Anyway, without the actual video I have to guess what the issue is…

Videos are encoded to optimize normal playback at the cost of speed of seeking. One of the techniques is to encode a keyframe and then apply differences to that keyframe. This compression technique save lots of space and works fine for forward playback. But if you seek to a random spot it usually lands on a frame that is only represented by differences from the previous frame. To actually seek to that exact spot the player has to look around for a full key frame and then apply changes until the exact spot is reached. The person encoding a video gets to decide how often they want to emit keyframes so depending upon the how the video was encoded there might be a significant lag while the player decodes frames to get to the exact spot. To avoid this lag and make seeking responsive many players seek to a nearby keyframe instead of the exact spot.

For audio/video support IINA uses a library from the [mpv](https://github.com/mpv-player/mpv) project. If you bring up IINA's settings and go to the `Key Bindings` tab and then check `Display raw values` you can see the commands being sent to mpv for the right and left arrow keys, `seek 5` and `seek -5`. From the mpv manual section on the [seek](https://mpv.io/manual/stable/#command-interface-seek-%3Ctarget%3E-[%3Cflags%3E]) command:
> **seek \<target> \[\<flags>]**
> …
> By default, keyframes is used for relative, relative-percent, and absolute-percent seeks, while exact is used for absolute seeks.

See the link for the full information on this command, but the takeaway is that mpv does not seek to the exact position unless the `exact` flag is specified.

To configure the right and left keys to exactly seek 5 seconds, click on the `Duplicate…` button to create a new set of key bindings that you can modify. Then with `Display raw values` checked, look for the RIGHT and LEFT keys. Change the values to be:
- `seek 5 exact`
- `seek -5 exact`

Then test to see if exact seeking corrects the issue.

**Comment by BloodmoonCyberFox** [comment]
it is not working, well, obviously from your reply, you already know it is problem how the video is encoded...

seems the way is to switch to other player..

**Reply by saagarjha** [reply]
We cannot investigate this issue without a copy of the video you're seeing problems with. While a screen recording is helpful, it only helps us confirm that there is a problem and make guesses as to where the issue may be. It is not enough information for us to fix the issue. Notably, even if we did know how to fix the problem, we would be unable to check that it resolved the issue. If you are not willing to share the video publicly, you can also send it to one of the developers by email (e.g. saagar@saagarjha.com for me). Otherwise we'll keep the issue in mind but do not have much we can do to help you with here. If another video player plays the video correctly, it might be a better choice for you :)

**Comment by BloodmoonCyberFox** [comment]
anyway, the only problem happens only on iPhone videos that I captured...as
for others, they are fine

**Comment by low-batt** [comment]
Thank you for the report that this occurs with iPhone videos. That helps to confirm what we know about this defect.

My speculation above regarding the nature of the problem has proven to be wrong.

Other IINA users reported this as well in issue #4502. I spent a lot of time searching for a video that reproduced this problem as seeking was working fine for me with all the videos I tested. I finally found some that reproduced the problem on a site that reviewed the iPhone 13. Turns out my iPhone is too old and does not generate videos that show the problem
 
I have confirmed this is working properly in IINA 1.3.1. For audio/video support IINA uses a library from the [mpv](https://github.com/mpv-player/mpv) project. For 1.3.2 the libraries IINA is built upon were upgraded. I have reported this problem to the mpv project in issue https://github.com/mpv-player/mpv/issues/12047. No fix as of yet, but they have reproduced the problem.

**Comment by BloodmoonCyberFox** [comment]
Mine is iPhone 14 Pro (heard that’s the same chip with iPhone 15) - not
sure if there is such development dependencies across softwares

**Comment by low-batt** [comment]
In general, the problem with digital video is its size. Over the years the way video is encoded has changed as the computer industry has developed ways to shrink the file size while preserving the video quality. But at the same time cameras have improved giving us features such as 4K resolution that have increased the size of video files, thus providing an additional reason to continue to develop new formats capable of further reducing the size of videos.

Any of the following can be a factor in how a video is encoded:
- Camera
- Processor
- Operating System
- Camera Settings

The computer industry attempts to hide these details from users. That can make it hard to find documentation discussing such technical details. In the case of Apple such details can sometimes be found in documents published by Apple Support, for example: [Using HEIF or HEVC media on Apple devices](https://support.apple.com/en-us/HT207022)

As for what is going on in the case at hand, we need to see what the mpv project uncovers as the root cause of the defect.

**Comment by BloodmoonCyberFox** [comment]
If you are talking the compression, encoding methods and decoding ways to
interpret these formats/files, it could possibly go to deeper level of
compilers to explain
